src/lib/edit_script.rs

```diff
@@ -0,0 +1,330 @@
+        }
+        let mut s1: Vec<NodeId> = vec![];
+        for child in w.children(&store.from_arena) {
+            from_in_order.remove(&child);
+            if new_mappings.contains_from(&child) {
+                let mapped = new_mappings.get_to(&child).unwrap();
+                if x.children(&store.to_arena).any(|n| n == mapped) {
+                    s1.push(child);
+                }
+            }
+        }
+        let s2: Vec<NodeId> = vec![];
+        for child in x.children(&store.to_arena) {
+            if new_mappings.contains_to(&child) {
+                let mapped = new_mappings.get_from(&child).unwrap();
+                if w.children(&store.from_arena).any(|n| n == mapped) {
+                    s1.push(child);
+                }
+            }
+        }
+        let lcs = self.lcss(store, &s1, &s2);
+        for &(from, to) in &lcs {
+            from_in_order.insert(from);
+            to_in_order.insert(to);
+        }
+        for a in &s1 {
+            for b in &s2 {
+                if new_mappings.contains_from(a) && new_mappings.get_to(a).unwrap() == *b &&
+                   !lcs.contains(&(*a, *b)) {
+                    let k = self.find_pos(store, *b, new_mappings, to_in_order);
+                    let mut mov = Move::new(*a, w, k);
+                    debug!("Edit script: MOV {} {:?} Parent: {} {:?}",
+                           store.from_arena[*a].label,
+                           store.from_arena[*a].value,
+                           store.to_arena[w].label,
+                           store.to_arena[w].value);
+                    script.push(mov);
+                    mov.apply(&mut store.from_arena)?;
+                    from_in_order.insert(*a);
+                    to_in_order.insert(*b);
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Find the position of node x in to_arena.
+    fn find_pos<T: Clone + Debug + Eq + 'static>(&self,
+                                                 store: &MappingStore<T>,
+                                                 x: NodeId,
+                                                 new_mappings: &TemporaryMappingStore,
+                                                 to_in_order: &HashSet<NodeId>)
+                                                 -> u16 {
+        debug!("find_pos({})", x);
+        if x.is_root(&store.to_arena) {
+            return 0;
+        }
+        let y = store.to_arena[x].parent().unwrap();
+        let siblings = y.children(&store.to_arena).collect::<Vec<NodeId>>();
+        for child in &siblings {
+            if to_in_order.contains(child) {
+                if x == *child {
+                    return 0;
+                } else {
+                    break;
+                }
+            }
+        }
+        let x_pos = x.get_child_position(&store.to_arena).unwrap();
+        let mut v: Option<NodeId> = None;
+        for (i, child) in x.children(&store.to_arena).enumerate() {
```

Comment:
You might prefer `x.children(&store.to_arena).take(x_pos)` or similar over the `if i >= x_pos` dance inside the loop?

Reply:
Hrm.

So this:

```rust
let v = x.children(&store.to_arena).enumerate().filter(|&(i, _)| i >= x_pos).nth(0).map(|(_, c)| c);
```

is apparently different to this:

```rust
let mut v2: Option<NodeId> = None;
for (i, child) in x.children(&store.to_arena).enumerate() {
    if to_in_order.contains(&child) {
        v2 = Some(child);
    }
    if i >= x_pos {
        break;
    }
}
```

Reply:
I was thinking something like:

```rust
for child in x.children(&store.to_arena).take(x_pos) {
    ...
}
```

Reply:
That doesn't do the same thing, I think. Let me have another go...

Reply:
Why doesn't it do the same thing? It says "take the first x_pos items", which is what the `break` clause is also enforcing.